dockstore-webservice/src/main/java/io/dockstore/webservice/resources/WorkflowResource.java

```diff
@@ -882,6 +902,347 @@ private void updateInfo(Workflow oldWorkflow, Workflow newWorkflow) {
+        // The Zenodo deposit type for Dockstore will always be SOFTWARE
+        depositMetadata.setUploadType(DepositMetadata.UploadTypeEnum.SOFTWARE);
+        // A metadata description is required for Zenodo
+        String description = workflow.getDescription();
+        String descriptionStr = (description == null || description.isEmpty()) ? "n/a" : workflow.getDescription();
+        depositMetadata.setDescription(descriptionStr);
+        depositMetadata.setVersion(workflowVersion.getName());
+
+        setMetadataKeywords(depositMetadata, workflow);
+
+        setMetadataRelatedIdentifiers(depositMetadata, workflow);
+
+        setMetadataCreator(depositMetadata, workflow);
+
+        setMetadataPublicationDate(depositMetadata, workflowVersion);
+
+        setMetadataCommunities(depositMetadata);
+    }
+
+    /**
+     * Provision files to upload to Zenodo
+     * @param zendoClient Zenodo api client
+     * @param returnDeposit Deposit object for the new version
+     * @param depositionID ID of Zendo deposit to which files will be attached
+     * @param workflow    workflow for which DOI is registered
+     * @param workflowVersion workflow version for which DOI is registered
+     */
+    private void provisionWorkflowVersionUploadFiles(ApiClient zendoClient, Deposit returnDeposit,
+            int depositionID, Workflow workflow, WorkflowVersion workflowVersion) {
+        // Creating a new version copies the files from the previous version
+        // We want to delete these since we will upload a new set of files
+        // if creating a completely new deposit this should not cause a problem
+        FilesApi filesApi = new FilesApi(zendoClient);
+        List<DepositionFile> originalFilesInNewVersion = returnDeposit.getFiles();
+        Iterator filesIter = originalFilesInNewVersion.iterator();
+        while (filesIter.hasNext()) {
+            String fileIdStr = ((DepositionFile)filesIter.next()).getId();
+            filesApi.deleteFile(depositionID, fileIdStr);
+        }
+
+        // Add workflow version source files as a zip to the DOI upload deposit
+        // Borrow code from getWorkflowZip
+        Set<SourceFile> sourceFiles = workflowVersion.getSourceFiles();
+        java.nio.file.Path path = Paths.get(workflowVersion.getWorkingDirectory());
+        if (sourceFiles == null || sourceFiles.size() == 0) {
+            LOG.warn("No source files found to zip when creating DOI");
+            throw new CustomWebApplicationException("No source files found to"
+                    + " upload when creating DOI. Zenodo requires at lease one file"
+                    + " to be uploaded in order to create a DOI.", HttpStatus.SC_BAD_REQUEST);
+        } else {
+            String fileName = workflow.getWorkflowPath().replaceAll("/", "-") + ".zip";
+            OutputStream outputStream;
+            try {
+                outputStream = new FileOutputStream(fileName);
```

Comment:
Does this file get deleted by the FilesAPI on line 1076? If not, you will need to delete it.

You should create this is in a temporary directory. Right now you're putting it in essentially a user-specified location, which can be dangerous.

Reply:
done

Reply:
https://ucsc-cgl.atlassian.net/browse/DOCK-823